src/main/java/com/arpnetworking/kairosdb/aggregators/MovingWindowAggregator.java

```diff
@@ -0,0 +1,294 @@
+
+            TimeUnit tu = m_sampling.getUnit();
+            switch (tu) {
+            case YEARS:
+                m_unitField = chronology.year();
+                break;
+            case MONTHS:
+                m_unitField = chronology.monthOfYear();
+                break;
+            case WEEKS:
+                m_unitField = chronology.weekOfWeekyear();
+                break;
+            case DAYS:
+                m_unitField = chronology.dayOfMonth();
+                break;
+            case HOURS:
+                m_unitField = chronology.hourOfDay();
+                break;
+            case MINUTES:
+                m_unitField = chronology.minuteOfHour();
+                break;
+            case SECONDS:
+                m_unitField = chronology.secondOfDay();
+                break;
+            default:
+                m_unitField = chronology.millisOfSecond();
+                break;
+            }
+        }
+
+        protected long getStartRange(long timestamp) {
+            long samplingValue = m_sampling.getValue();
+            long difference = m_unitField.getDifferenceAsLong(timestamp, m_startTime);
+            if (m_unitField.remainder(timestamp - m_startTime) > 0) {
+                difference += 1;
+            }
+            return m_unitField.add(m_startTime, difference + -1 * samplingValue);
+        }
+
+        protected long getEndRange(long timestamp) {
+            long difference = m_unitField.getDifferenceAsLong(timestamp, m_startTime);
+            return m_unitField.add(m_startTime, difference) - 1;
+        }
+
+        @Override
+        public DataPoint next() {
+            if (m_dpBuffer.isEmpty() || !m_dpIterator.hasNext()) {
+
+                // We calculate start and end ranges as the ranges may not be
+                // consecutive if data does not show up in each range.
+                long startRange; 
+
+                if (m_dpBuffer.isEmpty()) {
+                    // This is the first datapoint so front-load the sample window amount of data.
+                    startRange = m_startTime;
+                } else {
+                    startRange = getStartRange(currentDataPoint.getTimestamp());
+                }
+
+                long endRange = m_unitField.add(startRange, m_sampling.getValue());
+                m_currentRangeTime = endRange;
+
+                // Trim off any data that is too old
+                m_dpBuffer = new ConcurrentSkipListMap<>(m_dpBuffer.tailMap(startRange, false));
+
+                // Add data up until the end of the range
+                while (currentDataPoint != null && currentDataPoint.getTimestamp() <= endRange) {
+                    if (currentDataPoint.getTimestamp() > startRange) {
+                        m_dpBuffer.put(currentDataPoint.getTimestamp(), currentDataPoint);
+                    }
+                    if (hasNextInternal()) {
+                        currentDataPoint = nextInternal();
+                    }
```

Comment:
Missing `else: currentDataPoint = null;` or has next check in `while` condition?

Reply:
hasNextInternal sets currentDataPoint to null if 'false' as part of it's contract, setting it to null here would be redundant.

Reply:
Sorry, missed that, that's an interesting design.